consensus/propagation/have_wants.go

```diff
@@ -37,25 +36,28 @@ func (blockProp *Reactor) handleHaves(peer p2p.ID, haves *proptypes.HaveParts, b
 		blockProp.Logger.Error("peer not found", "peer", peer)
 		return
 	}
+
 	_, parts, fullReqs, has := blockProp.getAllState(height, round)
 	if !has {
 		// TODO disconnect from the peer
 		blockProp.Logger.Error("received part state for unknown proposal", "peer", peer, "height", height, "round", round)
 		return
 	}
 
-	blockProp.mtx.RLock()
-	defer blockProp.mtx.RUnlock()
+	p.Initialize(height, round, int(parts.Total()))
 
-	// Update the peer's haves.
-	p.SetHaves(height, round, haves)
+	bm, _ := p.GetHaves(height, round)
 
-	if parts.IsComplete() {
+	for _, pmd := range haves.Parts {
+		bm.SetIndex(int(pmd.Index), true)
+	}
+
+	if parts.Original().IsComplete() {
```

Comment:
why do we return? shouldn't we broadcast the haves we received?

if I understand correctly, this case can happen when:

- We receive a proposal
- We check the mempool and find all the necessary transactions locally
- Create the block locally
- Receive the haves from the proposer

Reply:
that makes sense!! we either should be broadcasting the haves as we get them or after we reconstruct. I think this is assuming that we are broadcasting haves after we reconstruct, but we aren't!